src/storage/src/source/kafka.rs

```diff
@@ -104,382 +98,423 @@ pub struct KafkaOffsetCommiter {
-                break;
-            }
-
-            let message = self.poll_from_next_queue();
-            attempts += 1;
+                // Poll the consumer once. We split the consumer's partitions out into separate
+                // queues and poll those individually, but it's still necessary to drive logic that
+                // consumes from rdkafka's internal event queue, such as statistics callbacks.
+                //
+                // Additionally, assigning topics and splitting them off into separate queues is
+                // not atomic, so we expect to see at least some messages to show up when polling
+                // the consumer directly.
+                if let Some(result) = reader.consumer.poll(Duration::from_secs(0)) {
+                    match result {
+                        Err(e) => {
+                            let error = format!(
+                                "kafka error when polling consumer for source: {} topic: {} : {}",
+                                reader.source_name, reader.topic_name, e
+                            );
+                            let status = HealthStatusUpdate::from(HealthStatus::StalledWithError {
+                                error,
+                                hint: None,
+                            });
+                            health_output.give(&health_cap, status).await;
+                        }
+                        Ok(message) => {
+                            let (message, ts) =
+                                construct_source_message(&message, reader.include_headers);
+                            if let Some((msg, time, diff)) = reader.handle_message(message, ts) {
+                                let pid = time.partition().unwrap();
+                                let part_cap = &reader.partition_capabilities[pid];
+                                data_output.give(part_cap, (Ok(msg), time, diff)).await;
+                            }
+                        }
+                    }
+                }
 
-            match message {
-                Ok(Some((message, ts))) => {
-                    next_message = self.handle_message(Ok(message), ts);
+                reader.update_stats();
+
+                // Take the consumers temporarily to get around borrow checker errors
+                let mut consumers = std::mem::take(&mut reader.partition_consumers);
+                for consumer in consumers.iter_mut() {
+                    while let Some(message) = consumer.get_next_message().transpose() {
+                        match message.map(|(msg, ts)| reader.handle_message(msg, ts)) {
+                            Ok(Some((msg, time, diff))) => {
+                                let pid = time.partition().unwrap();
+                                let part_cap = &reader.partition_capabilities[pid];
+                                data_output.give(part_cap, (Ok(msg), time, diff)).await;
```

Comment:
Might be worth a comment saying something like:
"Note that we produce definite data with respect to the `time` part of the `Collection` triple, but when resuming, data may be grouped at different timely timestamps"

Reply:
That sounds like something that belongs in the document describing what definite means. I find comments that explain concepts instead of what this particular piece of code is doing to be noisy and I much prefer that the code has enough keywords for whoever is reading it to look them up. In this case I have included documentation in the `SourceRender::render` method that mentions definite collections.

I will update that doc comment to have a link to the design doc

Reply:
thanks!!